common/Obelisk/OAuth/Config.hs

```diff
@@ -0,0 +1,215 @@
+type OAuthConfigPrivate r = OAuthConfig OAuthClientSecret r
+
+-- | `OAuthConfig` without client secret - for frontend.
+type OAuthConfigPublic r = OAuthConfig () r
+
+
+-- | Create a `OAuthConfigPublic` by reading values through "Obelisk.ExecutableConfig"
+--
+--   `_oAuthConfig_clientId` will be read from config/common/oauth/${OAuthProvider}/client-id
+--
+--   `_oAuthConfig_providerUri` will be read from config/common/oauth/${OAuthProvider}/uri
+--
+--   `fst` `_oAuthConfig_redirectUri` will be read from config/common/route
+--
+--   `_oAuthConfig_scope` will be read from
+--   config/common/oauth/${OAuthProvider}/scope (as space delimited list)
+--
+--   Where `OAuthProvider` in the above paths is coming from the first parameter of this function.
+--
+--   The returned `OAuthConfig` will have `_oAuthConfig_responseType` set to
+--   `AuthorizationResponseType_Code`.
+getOAuthConfigPublic
+  :: MonadIO m
+  => OAuthProvider -- ^ Some name describing the OAuth service to use. (E.g. `OAuthProvider "github"`)
+  -> Maybe (r (R OAuthRoute))
+  -> m (OAuthConfigPublic r)
+getOAuthConfigPublic p@(OAuthProvider provider) mr = liftIO $ do
+  providerUri <- getTextConfigRequired $ "config/common/oauth/" <> provider <> "/uri"
+  mRedirectBase <- getTextConfig $ "config/common/route"
+  mScope <- getTextConfig $ "config/common/oauth" <> provider <> "/scope"
+  clientId <- getTextConfigRequired $ "config/common/oauth/" <> provider <> "/client-id"
+  pure $ OAuthConfig
+    { _oAuthConfig_responseType = AuthorizationResponseType_Code
+    , _oAuthConfig_provider = p
+    , _oAuthConfig_providerUri = providerUri
+    , _oAuthConfig_redirectUri = (,) <$> mRedirectBase <*> mr
+    , _oAuthConfig_scope = maybe [] T.words mScope
+    , _oAuthConfig_clientId = OAuthClientId clientId
+    , _oAuthConfig_clientSecret = ()
+    }
+
+
+-- | Builds upon `getOAuthConfigPublic`, but also initializes `_oAuthConfig_clientSecret`.
+--
+--   In addition `_oAutConfig_providerUri` will be read from
+--   "config/backend/oauth/${OAuthProvider}/uri" instead of
+--   "common/oauth/${OAuthProvider}/uri", so the resulting config is
+--   suitable for the actual token request.
+--
+--   `_oAuthConfig_clientSecret` will be read from
+--   config/backend/oauth/${OAuthProvider}/client-secret
+--
+getOAuthConfigPrivate
+  :: MonadIO m
+  => OAuthProvider
+  -> Maybe (r (R OAuthRoute))
+  -> m (OAuthConfigPrivate r)
+getOAuthConfigPrivate p@(OAuthProvider provider) mr = do
+  pubCfg <- getOAuthConfigPublic p mr
+
+  providerUri <- liftIO $ getTextConfigRequired $
+    "config/backend/oauth/" <> provider <> "/uri"
+
+  secret <- liftIO $ getTextConfigRequired $
+    "config/backend/oauth/" <> provider <> "/client-secret"
+
+  pure $ pubCfg
+    { _oAuthConfig_providerUri = providerUri
+    , _oAuthConfig_clientSecret = OAuthClientSecret secret
+    }
+
+-- Internal helper functions:
+
+
+-- | Retrieve configuration and fail with exception if it does not exist.
+getTextConfigRequired
```

Comment:
This function shouldn't be defined in this package.

Reply:
It is not exported - just used internally.